### PR TITLE
prediction withdrawal

### PR DESCRIPTION
**Pull request by lsabor**
Initial implementation of prediction withdrawal

Core:
- [x] endpoint creation (/questions/withdraw/)
- [x] backend view
- [x] core backend logic
- [x] add api documentation for withdraw endpoint
- [x] add FAQ on withdrawal
- [x] button (binary)
- [x] button (multiple choice)
- [x] button (continuous)
- [x] button (group)
- [x] button (conditional)
- [x] timeline "x" (binary)
- [x] timeline "x" (multiple choice)
- [x] timeline "x" (continuous)
- [x] timeline "x" (group)
- [x] timeline "x" (conditional)
- [x] cp display when there is a gap (binary)
- [x] cp display when there is a gap  (multiple choice)
- [x] cp display when there is a gap (continuous)
- [x] cp display when there is a gap  (group)
- [x] cp display when there is a gap  (conditional)
- [x] tile display & chip (binary)
- [x] tile display & chip  (multiple choice)
- [x] tile display & chip (continuous)
- [x] tile display & chip  (group)
- [x] tile display & chip  (conditional)
- [x] cp display on forecast input when there is a gap (binary)
- [x] cp display on forecast input when there is a gap  (multiple choice)
- [x] cp display on forecast input when there is a gap (continuous)
- [x] cp display on forecast input when there is a gap  (group)
- [x] cp display on forecast input when there is a gap  (conditional)
- [x] make this play nice with "hide cp"

Testing:
- [x] button withdraws
- [x] scoring works as expected
- [x] can't withdraw if no active prediction (frontend)
- [x] unit tests - can't withdraw if no active prediction (backend)

Rollout:
- [x] Only allow withdrawal on posts in no monetary tournaments
- [x] Add notation to code for easy future full rollout ("Feature Flag: prediction-withdrawal")
- [x] Add disclaimer to FAQ withdrawal section

Other:
- [x] add a withdrawn filter - predicted filter should be "live prediction" + "withdrawn"
- [x] withdrawal should deactivate global notification
- [x] if no active prediction, cannot share prediction in comments

Timeline graphs:
image
image
image

User 1 (withdrew a few times, majority of time had no forecast)
image
User 2 (forecast once and kept their forecast)
image

When there is no CP, break the graph
image
group questions:
image

### CODE REVIEW

**Comment by lsabor** [comment]
This is ready for review. I've launched in in dev.metaculus.com, and asked for QA from product on slack

Code is ready for review as well. It's not thoroughly polished, but I think a smoother implementation would require some graph refactoring which can come another time.

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx:102]: The plan is to address this TODO in a separate PR?

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx:97]: Same note here, probably makes sense to use `prevForecastValue`

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx:1]: I have the same notes on `withdrawError`, `prevForecastValue`, and a TODO for this and other forecast makers.

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx
Comment (lsabor) [reply front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx:102]: Yes. It's a future-proofing feature (pre-registering prediction withdrawal).
I suppose I could just remove the comment.

File: front_end/src/components/charts/multiple_choice_chart.tsx
Comment (ncarazon) [inline front_end/src/components/charts/multiple_choice_chart.tsx:191]: That's a bad practice to define react components inside other component body. Let's move it outside (can be the same file right after the `buildChartData`).

Additionally, it would be better to rename this component to something like `PredictionSymbol` or something similar.

File: front_end/src/components/charts/multiple_choice_chart.tsx
Comment (ncarazon) [inline front_end/src/components/charts/multiple_choice_chart.tsx:203]: Let's remove non-null assertion so it wouldn't crash the app in case unexpected data is received. We could guard unexpected data and return `null` as a fallback:

```
  if (
    typeof x !== "number" ||
    typeof y !== "number" ||
    typeof size !== "number"
  ) {
    return null;
  }
  ```

File: front_end/src/components/charts/numeric_chart.tsx
Comment (ncarazon) [inline front_end/src/components/charts/numeric_chart.tsx:181]: Do we actually have to run it on every cursor change event? It may cause performance issues. I think we could move it into a separate variable and additionally, utilize `useMemo` hook to reduce variable recalculations:

```
  const timestamps = useMemo(
    () =>
      uniq([
        ...aggregation.history.map((f) => f.start_time),
        ...aggregation.history.map((f) => f.end_time ?? f.start_time),
        actualCloseTime ?? Date.now() / 1000,
      ]).sort((a, b) => a - b),
    [actualCloseTime, aggregation.history]
  );
  
  ...
  
        onCursorChange={(value: CursorCoordinatesPropType) => {
        if (typeof value === "number" && onCursorChange) {
          onCursorChange(
            timestamps[timestamps.findIndex((t) => t > value) - 1]
          );
        }
      }}
  
  ```

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:666]: let's not use non-null assertion and gracefully handle nullable state

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:417]: let's not use non-null assertion and gracefully handle nullable state

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:426]: let's not use non-null assertion and gracefully handle nullable state

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:615]: let's not use non-null assertion and gracefully handle nullable state

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:636]: A bunch of code here is duplicated in `generateChoiceItemsFromGroupQuestions`. Can we reduce code duplication by introducing a reusable helper?

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:691]: these comments can be removed?

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:724]: can be removed

File: front_end/src/utils/charts.ts
Comment (ncarazon) [inline front_end/src/utils/charts.ts:841]: Let's use `QuestionType.Binary` instead of regular string

File: front_end/src/app/(main)/questions/[id]/components/detailed_question_card/multiple_choice_chart_card.tsx
Comment (lsabor) [reply front_end/src/app/(main)/questions/[id]/components/detailed_question_card/multiple_choice_chart_card.tsx:64]: Makes sense. also applying to next line

File: front_end/src/components/charts/numeric_chart.tsx
Comment (lsabor) [reply front_end/src/components/charts/numeric_chart.tsx:181]: Great point!

File: front_end/src/utils/charts.ts
Comment (lsabor) [reply front_end/src/utils/charts.ts:277]: Yes definitely

File: front_end/src/utils/charts.ts
Comment (lsabor) [reply front_end/src/utils/charts.ts:636]: I gave it a try, but they are different enough that factoring out to a helper function won't help very much.
I have given it more thought and I actually think it might be worth making it so the Question my_forecasts and aggregations.<aggregation_name> histories are already in this format, meaning no data mangling would be required in the front end.

I think due to the intent on more serious refactoring later, as well as my genuine attempt at refactoring now not working out, I'm going to keep it this way for the time being.

File: front_end/src/app/(main)/questions/helpers/filters.ts
Comment (lsabor) [inline front_end/src/app/(main)/questions/helpers/filters.ts:273]: @ncarazon I could use your help here.
I wanted to add 2 additional options - one for when a user has prediction and their prediction is still active, and another when they withdrew their most recent forecast.
The URL params are set up on the backend to support things like: `?forecaster_id=12345&withdrawn=true`

But I can't quite figure out how to do that with the filter components we have available. I think it'll take you 1/10th the time it would take me. Could you help me out?

Comment (ncarazon) [reply front_end/src/app/(main)/questions/helpers/filters.ts]: Pushed a commit. The solution is a bit hacky, but in order to properly support multiple filters being applied per a single option in the popup we would need to do a refactor of the component and adjust it in a bunch of places, which will take some time. So I've left a TODO to consider refactoring when there is time

**Comment by lsabor** [comment]
@ncarazon Ready for second review

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:69]: Q: We can't use `extractPrevBinaryForecastValue` anymore? 
Also, looks like `prevYesForecast` and `prevNoForecast` are not used anymore (actually used for sending GA events, but I assume it should be replaced with `prevYesForecastValue` and `prevNoForecastValue`)

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:29]: There is eslint error for this import:

image

Could you, please, adjust your editor to automatically run eslint and fix error when needed? Additionally, could you please run linter and make sure all errors are fixed before merging?

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:275]: `questionOptions` is react state and it shouldn't be mutated. You should use `useState` for updating it's value. For reference, see how this is done in submission callback:

```
    setQuestionOptions((prev) =>
      prev.map((prevChoice) => ({ ...prevChoice, isDirty: false }))
    );
```

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:281]: Q: Why do we have a different logic for handling errors when executing withdraw action compared to predict action? These actions seem to return the same payload in catch block.

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:368]: nit: this fragment is redundant

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx:73]: Similar notes about `extractPrevNumericForecastValue` and `prevYesForecast`, `prevNoForecast` props

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx:337]: Same note about mutation `questionOptions` state

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_continuous.tsx:479]: redundant fragment

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_binary.tsx:147]: redundant fragment

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx:79]: Similar note as for previous forecast makers. We have prev forecast pasted through props, extracted prev forecast value (previously BE returned `any` object, so this was done to ensure we're working with valid data on FE) and `previousForecast`. I think this should be normalized to have a single source for previous forecast

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx
Comment (ncarazon) [inline front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx:203]: redundant fragment

File: posts/models.py
Comment (hlbmtc) [inline posts/models.py:168]: As far as I recall, when you submit a forecast, we always set the `end_time` of the previous forecast to today's date, while your new forecast always has `end_time=None`. Could there ever be a situation where a user has only two forecasts, both with `end_time != None`? If so, why?

File: posts/services/feed.py
Comment (hlbmtc) [inline posts/services/feed.py:171]: I'd raplaced it with a single:
```
if withdrawn is not None:
            qs = qs.annotate_has_active_forecast(forecaster_id).filter(
                has_active_forecast=not withdrawn
            )
```

File: posts/services/feed.py
Comment (hlbmtc) [inline posts/services/feed.py:221]: unnecessary parenthesis

File: questions/services.py
Comment (hlbmtc) [inline questions/services.py:632]: I think it's probably makes sense to move this block here (in the bottom of the `after_forecast_actions`):

```python
# Running forecast post triggers
    for post in posts:
        # There may be situations where async jobs from `create_forecast` complete after
        # `run_on_post_forecast` is triggered. To maintain the correct sequence of execution,
        # we need to ensure that `run_on_post_forecast` runs only after all forecasts have been processed.
        #
        # As a temporary solution, we introduce a 10-second delay before execution
        # to ensure all forecasts are processed.
        run_on_post_forecast.send_with_options(args=(post.id,), delay=10_000)
```

This will reduce code duplication

File: questions/views.py
Comment (hlbmtc) [inline questions/views.py:154]: Replace all of these checks with just `serializer.is_valid(raise_exception=True)`. This will raise ValidationError in case of wrong data and will be automatically converted to the appropriate JSON response by Django

File: questions/services.py
Comment (hlbmtc) [inline questions/services.py:700]: I see we already performed such checks in `bulk_withdraw_forecasts_api_view `

File: tests/unit/test_questions/test_views.py
Comment (hlbmtc) [inline tests/unit/test_questions/test_views.py:10]: No tests

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx
Comment (lsabor) [reply front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_conditional/forecast_maker_conditional_binary.tsx:281]: Totally fine point. I don't remember why but I mutated it at some point and then never returned it back. I think it was since I was fiddling with the different error types on the backend and how they effect returned error statements.
Reverting to the predict event format.

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx
Comment (lsabor) [reply front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_continuous.tsx:79]: That was specifically for slider values.
the issue with using slider values is that not all continuous forecasts have slider values. And in the future, we would want the storage of the "input method" to be more options than just slider values.
Unfortunately, my sidestepping of using the previously used `prevForecast` isn't complete, and probably should have been done as a separate PR entirely, but alas, I'm in quite deep in this PR, so I'll just have to integrate abandoning the method of passing around slider values in place of the full record of the lastest user (or aggregation) forecast.

File: front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_multiple_choice.tsx
Comment (lsabor) [reply front_end/src/app/(main)/questions/[id]/components/forecast_maker/forecast_maker_question/forecast_maker_multiple_choice.tsx:74]: `activeUserForecast` is now an input to `generateChoiceOptions` creating `choicesForecasts`, so this is the single source I'd say.
I could move remaining checks to check against `choicesForecasts.forecast` existence down the line, but I find the code more legible to check for the existence of `activeUserForecast`, so my preference would be to keep it this way.

File: posts/models.py
Comment (lsabor) [reply posts/models.py:168]: Yes, withdrawal is the exact time when you'd have your most recent foreacst with a set end_time.
So this checks if there is any active forecast - meaning one with end_time of None.
If the user withdrew, then the end time of their final forecast was set to the moment of withdrawal.

File: questions/services.py
Comment (lsabor) [reply questions/services.py:700]: This is true.
The duplication is in `bulk_create_forecasts_api_view`, which I was copying.
I'll remove duplication from both.

**Comment by lsabor** [comment]
I've added what I'm considering to be the final touches.
Ready for final inspection.

**Review by ncarazon — APPROVED** [review]
LGTM!

File: tests/unit/test_questions/conftest.py
Comment (hlbmtc) [inline tests/unit/test_questions/conftest.py:44]: A small nit: `create_question` populates Question object which does not have `post` field, so `post=post` makes not sense. You can move post creation here: `post = factory_post(question=question)`

Comment (lsabor) [reply tests/unit/test_questions/conftest.py:44]: Make sense. Ok!

**Pull request merged**
